cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -205,16 +220,58 @@ private void processNotification(AppMetadataStore appMetadataStore,
           return;
         }
         BasicThrowable cause = decodeBasicThrowable(properties.get(ProgramOptionConstants.PROGRAM_ERROR));
-        recordedStatus =
+        recordedRunRecord =
           appMetadataStore.recordProgramStop(programRunId, endTimeSecs, programRunStatus, cause, messageIdBytes);
         break;
       default:
         // This should not happen
         LOG.error("Unsupported program status {} for program {}, {}", programRunStatus, programRunId, notification);
         return;
     }
-    if (recordedStatus != null) {
-      publishRecordedStatus(notification, programRunId, recordedStatus);
+    if (recordedRunRecord != null) {
+      publishRecordedStatus(notification, programRunId, recordedRunRecord.getStatus());
+      if (programRunStatus.isEndState()) {
+        // TODO: send a notification to deprovision the cluster
```

Comment:
Shouldn't send from here. It'll be the runtime monitor doing so (@CuriousVini piece). The runtime monitor is the one responsible for relaying messages from the runtime on the cloud back to master. It'll also check the program state and make sure it gets all necessary messages before instructing the runtime JVM process to shutdown, followed by a message send for deprovisioning.

Reply:
ok we'll have to coordinate then, depending on what gets merged first